File: tests/regression.py
#!/usr/bin/env python3
"""Fail-closed executable differential tests for the pinned Miranda oracle."""

from __future__ import annotations

import argparse
from dataclasses import dataclass
import hashlib
import os
from pathlib import Path
import re
import shutil
import stat
import subprocess
import sys
import tempfile


@dataclass(frozen=True)
class TestCase:
    name: str
    input: bytes
    script_content: bytes | None = None
    script_path: str | None = None
    support_files: tuple[str, ...] = ()


CASES = (
    TestCase("arithmetic_1_plus_2", b"1+2\n/q\n"),
    TestCase("factorial_10", b"product [1..10]\n/q\n"),
    TestCase("map_double", b"map (2*) [1..5]\n/q\n"),
    TestCase(
        "big_integers",
        b"12345678901234567890 + 10\n2^80\n/q\n",
    ),
    TestCase(
        "lazy_lists_and_strings",
        b'take 5 [1..]\nreverse [1,2,3]\nzip2 [1,2,3] [4,5,6]\n"abc" ++ "def"\n/q\n',
    ),
    TestCase(
        "fibonacci_script",
        b"fib 10\n/q\n",
        script_path="miralib/ex/fib.m",
    ),
    TestCase(
        "user_defined_script",
        b"square 12\ntwice square 2\npairup 1 2\n/q\n",
        script_content=b"square x = x*x\ntwice f x = f (f x)\npairup x y = (x,y)\n",
    ),
    TestCase("hex_octal_literals", b"0xff\n0x1A2B3C\n0o777\n\n/q\n"),
    TestCase(
        "literate_script",
        b"square 5\ncube 3\n\n/q\n",
        script_content=(
            b"> square x = x * x\n> cube x  = x * x * x\n\n"
            b"Prose below the code lines is not Miranda source.\n"
        ),
    ),
    TestCase(
        "insert_directive",
        b"r\n\n/q\n",
        script_content=b'%insert "directive_insert_body.txt"\n\nr = inserted_val + 1\n',
        support_files=("tests/golden/directive_insert_body.txt",),
    ),
)


@dataclass(frozen=True)
class Outcome:
    stdout: bytes
    stderr: bytes
    returncode: int
    timed_out: bool
    files: tuple[tuple[str, str, int, str], ...]


CPU_STAT = re.compile(rb"(?m)^(\|\|.*,\s*cpu\s*=\s*)[0-9]+(?:\.[0-9]+)?$")


def comparable_stderr(value: bytes) -> bytes:
    """Mask only nondeterministic CPU time; every other stderr byte is exact."""
    return CPU_STAT.sub(rb"\1<TIME>", value)


def snapshot(root: Path) -> tuple[tuple[str, str, int, str], ...]:
    entries: list[tuple[str, str, int, str]] = []
    for path in sorted(root.rglob("*"), key=lambda value: value.relative_to(root).as_posix()):
        relative = path.relative_to(root).as_posix()
        info = path.lstat()
        mode = stat.S_IMODE(info.st_mode)
        if path.is_symlink():
            entries.append((relative, "symlink", mode, os.readlink(path)))
        elif path.is_dir():
            entries.append((relative, "directory", mode, ""))
        elif path.is_file():
            entries.append(
                (relative, "file", mode, hashlib.sha256(path.read_bytes()).hexdigest())
            )
        else:
            entries.append((relative, "other", mode, ""))
    return tuple(entries)


def prepare_case(root: Path, case: TestCase, repository: Path) -> Path | None:
    for support in case.support_files:
        source = repository / support
        shutil.copy2(source, root / source.name)
    if case.script_content is not None:
        script = root / "case.m"
        script.write_bytes(case.script_content)
        return script
    if case.script_path is not None:
        return repository / case.script_path
    return None


def run(
    binary: Path,
    library: Path,
    repository: Path,
    case: TestCase,
    timeout: float,
) -> Outcome:
    with tempfile.TemporaryDirectory(prefix=f"miracula-{case.name}-") as temp:
        workdir = Path(temp)
        script = prepare_case(workdir, case, repository)
        argv = [
            str(binary),
            "-lib",
            str(library),
            "-hush",
            "-count",
        ]
        if script is not None:
            argv.append(str(script))
        try:
            completed = subprocess.run(
                argv,
                cwd=workdir,
                input=case.input,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                timeout=timeout,
                check=False,
            )
            return Outcome(
                stdout=completed.stdout,
                stderr=completed.stderr,
                returncode=completed.returncode,
                timed_out=False,
                files=snapshot(workdir),
            )
        except subprocess.TimeoutExpired as exc:
            return Outcome(
                stdout=exc.stdout or b"",
                stderr=exc.stderr or b"",
                returncode=-1,
                timed_out=True,
                files=snapshot(workdir),
            )


def describe_returncode(code: int) -> str:
    if code < 0:
        return f"signal {-code}"
    return f"exit {code}"


def compare(case: TestCase, expected: Outcome, actual: Outcome) -> list[str]:
    differences: list[str] = []
    if expected.timed_out:
        differences.append("reference timed out; the oracle case is invalid")
    if actual.timed_out:
        differences.append("candidate timed out")
    if expected.returncode != actual.returncode:
        differences.append(
            "termination mismatch: reference "
            f"{describe_returncode(expected.returncode)}, candidate "
            f"{describe_returncode(actual.returncode)}"
        )
    if expected.stdout != actual.stdout:
        differences.append(
            f"stdout mismatch ({len(expected.stdout)} reference bytes, "
            f"{len(actual.stdout)} candidate bytes)"
        )
    expected_stderr = comparable_stderr(expected.stderr)
    actual_stderr = comparable_stderr(actual.stderr)
    if expected_stderr != actual_stderr:
        differences.append(
            f"stderr mismatch after CPU-time masking ({len(expected.stderr)} "
            f"reference bytes, {len(actual.stderr)} candidate bytes)"
        )
    if expected.files != actual.files:
        expected_set = set(expected.files)
        actual_set = set(actual.files)
        for entry in sorted(expected_set - actual_set):
            differences.append(f"filesystem missing/different: {entry}")
        for entry in sorted(actual_set - expected_set):
            differences.append(f"filesystem unexpected/different: {entry}")
    return differences


def existing_file(value: str) -> Path:
    path = Path(value).resolve()
    if not path.is_file():
        raise argparse.ArgumentTypeError(f"file does not exist: {value}")
    return path


def existing_dir(value: str) -> Path:
    path = Path(value).resolve()
    if not path.is_dir():
        raise argparse.ArgumentTypeError(f"directory does not exist: {value}")
    return path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--candidate", required=True, type=existing_file)
    parser.add_argument("--reference", required=True, type=existing_file)
    parser.add_argument("--library", required=True, type=existing_dir)
    parser.add_argument("--repository", type=existing_dir, default=Path.cwd())
    parser.add_argument("--timeout", type=float, default=10.0)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    executed = 0
    passed = 0
    failed = 0
    skipped = 0

    for case in CASES:
        executed += 1
        print(f"Running differential case: {case.name} ... ", end="", flush=True)
        reference = run(
            args.reference, args.library, args.repository, case, args.timeout
        )
        candidate = run(
            args.candidate, args.library, args.repository, case, args.timeout
        )
        differences = compare(case, reference, candidate)
        if differences:
            failed += 1
            print("FAIL")
            for difference in differences:
                print(f"  {difference}")
            if reference.stdout != candidate.stdout:
                print(f"  reference stdout: {reference.stdout!r}")
                print(f"  candidate stdout: {candidate.stdout!r}")
            if reference.stderr != candidate.stderr:
                print(f"  reference stderr: {reference.stderr!r}")
                print(f"  candidate stderr: {candidate.stderr!r}")
        else:
            passed += 1
            print("PASS")

    print(
        f"Differential summary: executed={executed} passed={passed} "
        f"failed={failed} skipped={skipped}"
    )
    if executed == 0 or skipped != 0 or failed != 0:
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/regression.zig
const std = @import("std");

const TestCase = struct {
    name: []const u8,
    input: []const u8,
    script_content: ?[]const u8 = null,
    script_path: ?[]const u8 = null,
};

const TEST_CASES = [_]TestCase{
    .{
        .name = "arithmetic_1_plus_2",
        .input = "1+2\n/q\n",
    },
    .{
        .name = "factorial_10",
        .input = "product [1..10]\n/q\n",
    },
    .{
        .name = "map_double",
        .input = "map (2*) [1..5]\n/q\n",
    },
    .{
        .name = "big_integers",
        .input = "12345678901234567890 + 10\n2^80\n/q\n",
    },
    .{
        .name = "lazy_lists_and_strings",
        .input = "take 5 [1..]\nreverse [1,2,3]\nzip2 [1,2,3] [4,5,6]\n\"abc\" ++ \"def\"\n/q\n",
    },
    .{
        .name = "fibonacci_script",
        .script_path = "miralib/ex/fib.m",
        .input = "fib 10\n/q\n",
    },
    .{
        .name = "user_defined_script",
        .script_content = "square x = x*x\ntwice f x = f (f x)\npairup x y = (x,y)\n",
        .input = "square 12\ntwice square 2\npairup 1 2\n/q\n",
    },
    // docs/GoReady.md Phase 0 step 4: differential coverage for the
    // surfaces Phase 1's native lexer/parser rewrite will touch. Mirrors
    // tests/golden/hex_oct_literals and tests/golden/literate_intro.lit.
    .{
        .name = "hex_octal_literals",
        .input = "0xff\n0x1A2B3C\n0o777\n\n/q\n",
    },
    .{
        .name = "literate_script",
        .script_content = "> square x = x * x\n> cube x  = x * x * x\n\nProse below the code lines is not Miranda source.\n",
        .input = "square 5\ncube 3\n\n/q\n",
    },
    // %insert is pure textual substitution in the legacy lexer (no AST node),
    // so it already works; resolves relative to reg_tmp.m's directory
    // (tests/golden/), reusing directive_insert_body.txt there.
    .{
        .name = "insert_directive",
        .script_content = "%insert \"directive_insert_body.txt\"\n\nr = inserted_val + 1\n",
        .input = "r\n\n/q\n",
    },
};

const Stats = struct {
    reductions: u64 = 0,
    cells_claimed: u64 = 0,
    no_of_gcs: u64 = 0,
};

fn parseStats(allocator: std.mem.Allocator, stderr_data: []const u8, stats: *Stats) ![]const u8 {
    var other_err: std.ArrayList([]const u8) = .empty;
    defer other_err.deinit(allocator);

    var lines = std.mem.splitScalar(u8, stderr_data, '\n');
    while (lines.next()) |line| {
        if (std.mem.startsWith(u8, line, "||")) {
            if (std.mem.indexOf(u8, line, "reductions =")) |idx| {
                const sub = line[idx + "reductions =".len ..];
                const trimmed = std.mem.trim(u8, sub, " \t\r\n");
                var parts = std.mem.splitAny(u8, trimmed, ", \t\r\n");
                stats.reductions = std.fmt.parseInt(u64, parts.first(), 10) catch 0;
            }
            if (std.mem.indexOf(u8, line, "cells claimed =")) |idx| {
                const sub = line[idx + "cells claimed =".len ..];
                const trimmed = std.mem.trim(u8, sub, " \t\r\n");
                var parts = std.mem.splitAny(u8, trimmed, ", \t\r\n");
                stats.cells_claimed = std.fmt.parseInt(u64, parts.first(), 10) catch 0;
            }
            if (std.mem.indexOf(u8, line, "no of gc's =")) |idx| {
                const sub = line[idx + "no of gc's =".len ..];
                const trimmed = std.mem.trim(u8, sub, " \t\r\n");
                var parts = std.mem.splitAny(u8, trimmed, ", \t\r\n");
                stats.no_of_gcs = std.fmt.parseInt(u64, parts.first(), 10) catch 0;
            }
        } else {
            if (!std.mem.startsWith(u8, line, "[TRACE]") and
                !std.mem.startsWith(u8, line, "[ALLOC]") and
                !std.mem.startsWith(u8, line, "[DIAG"))
            {
                try other_err.append(allocator, line);
            }
        }
    }

    var total_len: usize = 0;
    for (other_err.items) |line| {
        total_len += line.len;
    }
    if (other_err.items.len > 0) {
        total_len += other_err.items.len - 1;
    }

    const joined = try allocator.alloc(u8, total_len);
    errdefer allocator.free(joined);
    var offset: usize = 0;
    for (other_err.items, 0..) |line, idx| {
        std.mem.copyForwards(u8, joined[offset..], line);
        offset += line.len;
        if (idx < other_err.items.len - 1) {
            joined[offset] = '\n';
            offset += 1;
        }
    }
    return joined;
}

const RunResult = struct {
    stdout: []const u8,
    stderr: []const u8,
    returncode: i32,
};

fn runBinary(allocator: std.mem.Allocator, io: std.Io, binary_path: []const u8, tc: TestCase) !RunResult {
    const temp_path = "./tests/golden/reg_tmp.m";
    var has_temp = false;

    const argv_len = 7;
    var argv = try allocator.alloc([]const u8, argv_len);
    defer allocator.free(argv);

    var argc: usize = 0;
    argv[argc] = binary_path;
    argc += 1;
    argv[argc] = "-lib";
    argc += 1;
    argv[argc] = "miralib";
    argc += 1;
    argv[argc] = "-hush";
    argc += 1;
    argv[argc] = "-count";
    argc += 1;

    if (tc.script_content) |content| {
        const file = try std.Io.Dir.cwd().createFile(io, temp_path, .{});
        try file.writeStreamingAll(io, content);
        file.close(io);
        argv[argc] = temp_path;
        argc += 1;
        has_temp = true;
    } else if (tc.script_path) |path| {
        argv[argc] = path;
        argc += 1;
    }
    defer if (has_temp) {
        std.Io.Dir.cwd().deleteFile(io, temp_path) catch {};
    };

    var child = try std.process.spawn(io, .{
        .argv = argv[0..argc],
        .stdin = .pipe,
        .stdout = .pipe,
        .stderr = .pipe,
    });
    errdefer child.kill(io);

    child.stdin.?.writeStreamingAll(io, tc.input) catch {};
    child.stdin.?.close(io);
    child.stdin = null;

    var multi_reader_buffer: std.Io.File.MultiReader.Buffer(2) = undefined;
    var multi_reader: std.Io.File.MultiReader = undefined;
    multi_reader.init(allocator, io, multi_reader_buffer.toStreams(), &.{ child.stdout.?, child.stderr.? });
    defer multi_reader.deinit();

    const timeout = std.Io.Timeout{ .duration = .{ .raw = std.Io.Duration.fromSeconds(10), .clock = .real } };
    while (multi_reader.fill(64, timeout)) |_| {} else |err| switch (err) {
        error.EndOfStream => {},
        error.Timeout => {
            child.kill(io);
            _ = try child.wait(io);
            return error.Timeout;
        },
        else => |e| return e,
    }

    const term = try child.wait(io);
    const returncode = switch (term) {
        .exited => |code| @as(i32, @intCast(code)),
        else => -1,
    };

    const stdout_text = try multi_reader.toOwnedSlice(0);
    const stderr_text = try multi_reader.toOwnedSlice(1);

    return RunResult{
        .stdout = stdout_text,
        .stderr = stderr_text,
        .returncode = returncode,
    };
}

pub fn main(ctx: std.process.Init) !void {
    var gpa = std.heap.DebugAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const original_bin = "./mira_original";
    const refactored_bin = "./zig-out/bin/mira";

    _ = std.Io.Dir.cwd().statFile(ctx.io, original_bin, .{}) catch {
        std.debug.print("Original binary {s} not found, skipping differential regression checks.\n", .{original_bin});
        std.process.exit(0);
    };

    _ = std.Io.Dir.cwd().statFile(ctx.io, refactored_bin, .{}) catch {
        std.debug.print("Error: Refactored binary {s} not found.\n", .{refactored_bin});
        std.process.exit(1);
    };

    var failed = false;

    for (TEST_CASES) |tc| {
        std.debug.print("Running test: {s} ... ", .{tc.name});

        const orig_res = try runBinary(allocator, ctx.io, original_bin, tc);
        defer allocator.free(orig_res.stdout);
        defer allocator.free(orig_res.stderr);

        const ref_res = try runBinary(allocator, ctx.io, refactored_bin, tc);
        defer allocator.free(ref_res.stdout);
        defer allocator.free(ref_res.stderr);

        if (orig_res.returncode != ref_res.returncode) {
            std.debug.print("FAIL (Exit code mismatch)\n", .{});
            std.debug.print("  Original: {d}\n", .{orig_res.returncode});
            std.debug.print("  Refactored: {d}\n", .{ref_res.returncode});
            failed = true;
            continue;
        }

        if (!std.mem.eql(u8, orig_res.stdout, ref_res.stdout)) {
            std.debug.print("FAIL (Stdout mismatch)\n", .{});
            std.debug.print("--- Original Stdout ---\n{s}\n", .{orig_res.stdout});
            std.debug.print("--- Refactored Stdout ---\n{s}\n", .{ref_res.stdout});
            failed = true;
            continue;
        }

        var orig_stats = Stats{};
        var ref_stats = Stats{};

        const orig_other_err = try parseStats(allocator, orig_res.stderr, &orig_stats);
        defer allocator.free(orig_other_err);
        const ref_other_err = try parseStats(allocator, ref_res.stderr, &ref_stats);
        defer allocator.free(ref_other_err);

        if (!std.mem.eql(u8, orig_other_err, ref_other_err)) {
            std.debug.print("FAIL (Stderr mismatch, excluding stats)\n", .{});
            std.debug.print("--- Original Stderr ---\n{s}\n", .{orig_other_err});
            std.debug.print("--- Refactored Stderr ---\n{s}\n", .{ref_other_err});
            failed = true;
            continue;
        }

        if (orig_stats.reductions != ref_stats.reductions or
            orig_stats.cells_claimed != ref_stats.cells_claimed or
            orig_stats.no_of_gcs != ref_stats.no_of_gcs)
        {
            std.debug.print("FAIL (Stats mismatch: orig_red={d}, ref_red={d})\n", .{ orig_stats.reductions, ref_stats.reductions });
            failed = true;
            continue;
        }

        std.debug.print("PASS\n", .{});
    }

    if (failed) {
        std.process.exit(1);
    } else {
        std.debug.print("All verification tests passed successfully!\n", .{});
        std.process.exit(0);
    }
}
